src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,526 @@
+  /// Set a resource to the given value.
+  /// NOTE: if the new value is 0, the resource will be removed.
+  ResourceRequest &Set(ResourceID resource_id, FixedPoint value) {
+    if (value == 0) {
+      if (IsPredefinedResource(resource_id)) {
+        predefined_resources_[resource_id.ToInt()] = 0;
+      } else {
+        custom_resources_.erase(resource_id.ToInt());
+      }
+    } else {
+      auto ptr = GetPointer(resource_id);
+      if (ptr == nullptr) {
+        custom_resources_[resource_id.ToInt()] = value;
+      } else {
+        *ptr = value;
+      }
+    }
+    return *this;
+  }
+
+  /// Check whether a particular resource exist.
+  bool Has(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    return ptr != nullptr && *ptr != 0;
+  }
+
+  /// Clear the whole set.
+  void Clear() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      predefined_resources_[i] = 0;
+    }
+    custom_resources_.clear();
+  }
+
+  /// Cap the resource values in this set by those in another set.
+  void Cap(const ResourceRequest &other) {
+    for (auto &resource_id : ResourceIds()) {
+      auto this_value = Get(resource_id);
+      auto other_value = other.Get(resource_id);
+      if (this_value > other_value) {
+        Set(resource_id, other_value);
+      }
+    }
+  }
+
+  /// Remove the negative values in this set.
+  void RemoveNegative() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] < 0) {
+        predefined_resources_[i] = 0;
+      }
+    }
+    for (auto it = custom_resources_.begin(); it != custom_resources_.end();) {
+      if (it->second < 0) {
+        custom_resources_.erase(it++);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  /// Return the number of resources in this set.
+  size_t Size() const {
+    size_t size = custom_resources_.size();
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      if (predefined_resources_[i] != 0) {
+        size++;
+      }
+    }
+    return size;
+  }
+
+  /// Return true if this set is empty.
+  bool IsEmpty() const { return Size() == 0; }
+
+  /// Return a set that contains all resource ids in this set.
+  absl::flat_hash_set<ResourceID> ResourceIds() const {
+    absl::flat_hash_set<ResourceID> res;
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] != 0) {
+        res.insert(ResourceID(i));
+      }
+    }
+    for (auto &entry : custom_resources_) {
+      res.insert(ResourceID(entry.first));
+    }
+    return res;
+  }
+
+  /// Return a map from the resource ids to the values.
+  absl::flat_hash_map<ResourceID, FixedPoint> ToMap() const {
+    absl::flat_hash_map<ResourceID, FixedPoint> res;
+    for (auto &resource_id : ResourceIds()) {
+      res.emplace(resource_id, Get(resource_id));
+    }
+    return res;
+  }
+
+  ResourceRequest &operator=(const ResourceRequest &other) {
```

Comment:
This is auto generated by compiler?

Reply:
Are you suggesting "= default"?

Reply:
You don't even need "=default". We can just omit this function and compiler will generate one for us I think. But "=default" is also fine to me.